### DISCUSSION TITLE
C++ Lock-free Atomic Shared Pointer | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#161] [Announcements]
https://sf-zhou.github.io/programming/atomic_shared_ptr.html

**Comment by orrorcol** [comment]
``` c++
ReferenceCount *Ref() {
    ++cnt_; // 1
    return this;
  }

  void Deref() {
    if (--cnt_ == 0) { // 2
      delete this; // 3
    }
  }
```

这个不是线程安全的吧, 可能的执行循序有 2 -> 1 -> 3

**Comment by SF-Zhou** [comment]
> @uestc-lfs
> ``` c++
> ReferenceCount *Ref() {
>     ++cnt_; // 1
>     return this;
>   }
> 
>   void Deref() {
>     if (--cnt_ == 0) { // 2
>       delete this; // 3
>     }
>   }
> ```
> 
> 这个不是线程安全的吧, 可能的执行循序有 2 -> 1 -> 3

你说的这种情况确实会导致线程不安全，但在 SharedPtr 单线程写或多线程读场景下并不会出现。

**Comment by orrorcol** [comment]
哦哦，是

获取 Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Stephan14** [comment]
```return reinterpret_cast<ReferenceCount<T> *>(local & (-1ull >> 16))```
有点没有明白这行代码，这个是为了把local的前16为置为0，然后获取ReferenceCount<T>的地址的指针吗？但是-1ull >> 16的16进制是ffffffffffff

**Reply by SF-Zhou** [reply]
前 16 位置0，留下来的 48 位就是指针

**Comment by aprilweet** [comment]
```cpp 
ReferenceCount<T> *Acquire() {
    uint64_t local = 0;
    do {
      local = rc_.load();
    } while (!rc_.compare_exchange_weak(local, local - (1ull << 48)));
    return reinterpret_cast<ReferenceCount<T> *>(local & (-1ull >> 16));
  }
```
这里应该是要加1吧，而不是减1。也就是：
```cpp
  } while (!rc_.compare_exchange_weak(local, local + (1ull << 48)));
```

**Reply by SF-Zhou** [reply]
代码没有问题，建议你跑下代码验证试试

**Reply by wangyu-xreal** [reply]
这里减的是AtomicSharedPtr x本地的引用计数， SharedPtr<A> b在析构时会减引用计数， 所以这块x本地计数减1，之后AtomicSharedPtr x析构时引用计数减8191就正确了；

前面 x.Store(a); 时引用计数会增加8192；
